File: script/make/protocol/lua/AccountProtocol.lua
AccountProtocol = {}

function AccountProtocol.encode(offset, protocol, data)
    if protocol == 10000 then
        local table = {}

        return table
    elseif protocol == 10001 then
        local table = {}

        -- 服务器ID
        table[offset] = string.pack(">I2", data["serverId"])
        offset = offset + 1
        -- 账户名
        table[offset] = string.pack(">s2", data["accountName"])
        offset = offset + 1
        return table
    elseif protocol == 10002 then
        local table = {}

        -- 角色名
        table[offset] = string.pack(">s2", data["roleName"])
        offset = offset + 1
        -- 服务器ID
        table[offset] = string.pack(">I2", data["serverId"])
        offset = offset + 1
        -- 账户名
        table[offset] = string.pack(">s2", data["accountName"])
        offset = offset + 1
        -- 性别
        table[offset] = string.pack(">I1", data["sex"])
        offset = offset + 1
        -- 职业
        table[offset] = string.pack(">I1", data["classes"])
        offset = offset + 1
        -- 渠道
        table[offset] = string.pack(">s2", data["channel"])
        offset = offset + 1
        -- 设备
        table[offset] = string.pack(">s2", data["deviceId"])
        offset = offset + 1
        -- mac地址
        table[offset] = string.pack(">s2", data["mac"])
        offset = offset + 1
        -- 设备类型
        table[offset] = string.pack(">s2", data["deviceType"])
        offset = offset + 1
        return table
    elseif protocol == 10003 then
        local table = {}

        -- 角色ID
        table[offset] = string.pack(">I8", data["roleId"])
        offset = offset + 1
        -- 角色名
        table[offset] = string.pack(">s2", data["roleName"])
        offset = offset + 1
        -- 服务器ID
        table[offset] = string.pack(">I2", data["serverId"])
        offset = offset + 1
        -- 账户名
        table[offset] = string.pack(">s2", data["accountName"])
        offset = offset + 1
        return table
    elseif protocol == 10004 then
        local table = {}

        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function AccountProtocol.decode(offset, protocol, bytes)
    if protocol == 10000 then
        -- 结果
        local data = string.unpack(">s2", bytes, offset)
        offset = offset + 2 + string.len(data)
        return data
    elseif protocol == 10001 then
        -- 
        -- 结果
        local dataResult = string.unpack(">s2", bytes, offset)
        offset = offset + 2 + string.len(dataResult)
        -- 角色名列表
        local dataList = {}
        local dataListLength = string.unpack(">I2", bytes, offset)
        offset = offset + 2
        for dataListIndex = 1, dataListLength do
            -- 
            -- 角色ID
            local dataListDataRoleId = string.unpack(">I8", bytes, offset)
            offset = offset + 8
            -- 角色名
            local dataListDataRoleName = string.unpack(">s2", bytes, offset)
            offset = offset + 2 + string.len(dataListDataRoleName)
            -- object
            local dataListData = {roleId = dataListDataRoleId, roleName = dataListDataRoleName}
            dataList[dataListIndex] = dataListData
        end
        -- object
        local data = {result = dataResult, list = dataList}
        return data
    elseif protocol == 10002 then
        -- 
        -- 结果
        local dataResult = string.unpack(">s2", bytes, offset)
        offset = offset + 2 + string.len(dataResult)
        -- 角色ID
        local dataRoleId = string.unpack(">I8", bytes, offset)
        offset = offset + 8
        -- 角色名
        local dataRoleName = string.unpack(">s2", bytes, offset)
        offset = offset + 2 + string.len(dataRoleName)
        -- object
        local data = {result = dataResult, roleId = dataRoleId, roleName = dataRoleName}
        return data
    elseif protocol == 10003 then
        -- 结果
        local data = string.unpack(">s2", bytes, offset)
        offset = offset + 2 + string.len(data)
        return data
    elseif protocol == 10004 then
        -- 结果
        local data = string.unpack(">s2", bytes, offset)
        offset = offset + 2 + string.len(data)
        return data
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

File: script/make/protocol/cs/AccountProtocol.cs
public static class AccountProtocol
{
    public static void Encode(System.Text.Encoding encoding, System.IO.BinaryWriter writer, System.UInt16 protocol, System.Collections.Generic.Dictionary<System.String, System.Object> data) 
    {
        switch (protocol) 
        {
            case 10000:
            {
                return;
            }
            case 10001:
            {
                // 服务器ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)(System.UInt16)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["serverId"]));
                // 账户名
                var accountNameBytes = encoding.GetBytes((System.String)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["accountName"]);
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)accountNameBytes.Length));
                writer.Write(accountNameBytes);
                return;
            }
            case 10002:
            {
                // 角色名
                var roleNameBytes = encoding.GetBytes((System.String)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["roleName"]);
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)roleNameBytes.Length));
                writer.Write(roleNameBytes);
                // 服务器ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)(System.UInt16)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["serverId"]));
                // 账户名
                var accountNameBytes = encoding.GetBytes((System.String)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["accountName"]);
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)accountNameBytes.Length));
                writer.Write(accountNameBytes);
                // 性别
                writer.Write((System.Byte)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["sex"]);
                // 职业
                writer.Write((System.Byte)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["classes"]);
                // 渠道
                var channelBytes = encoding.GetBytes((System.String)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["channel"]);
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)channelBytes.Length));
                writer.Write(channelBytes);
                // 设备
                var deviceIdBytes = encoding.GetBytes((System.String)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["deviceId"]);
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)deviceIdBytes.Length));
                writer.Write(deviceIdBytes);
                // mac地址
                var macBytes = encoding.GetBytes((System.String)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["mac"]);
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)macBytes.Length));
                writer.Write(macBytes);
                // 设备类型
                var deviceTypeBytes = encoding.GetBytes((System.String)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["deviceType"]);
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)deviceTypeBytes.Length));
                writer.Write(deviceTypeBytes);
                return;
            }
            case 10003:
            {
                // 角色ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int64)(System.UInt64)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["roleId"]));
                // 角色名
                var roleNameBytes = encoding.GetBytes((System.String)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["roleName"]);
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)roleNameBytes.Length));
                writer.Write(roleNameBytes);
                // 服务器ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)(System.UInt16)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["serverId"]));
                // 账户名
                var accountNameBytes = encoding.GetBytes((System.String)((System.Collections.Generic.Dictionary<System.String, System.Object>)data["data"])["accountName"]);
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)accountNameBytes.Length));
                writer.Write(accountNameBytes);
                return;
            }
            case 10004:
            {
                return;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }

    public static System.Collections.Generic.Dictionary<System.String, System.Object> Decode(System.Text.Encoding encoding, System.IO.BinaryReader reader, System.UInt16 protocol) 
    {
        switch (protocol) 
        {
            case 10000:
            {
                // 结果
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = encoding.GetString(reader.ReadBytes(dataLength));
                return data;
            }
            case 10001:
            {
                // 
                // 结果
                var resultLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var result = encoding.GetString(reader.ReadBytes(resultLength));
                // 角色名列表
                var listLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var list = new System.Collections.Generic.List<System.Object>(listLength);
                while (listLength-- > 0)
                {
                    // 
                    // 角色ID
                    var roleId = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 角色名
                    var roleNameLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var roleName = encoding.GetString(reader.ReadBytes(roleNameLength));
                    // object
                    var listItem = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"roleId", roleId}, {"roleName", roleName}};
                    // add
                    list.Add(listItem);
                }
                // object
                var data = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"result", result}, {"list", list}};
                return data;
            }
            case 10002:
            {
                // 
                // 结果
                var resultLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var result = encoding.GetString(reader.ReadBytes(resultLength));
                // 角色ID
                var roleId = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                // 角色名
                var roleNameLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var roleName = encoding.GetString(reader.ReadBytes(roleNameLength));
                // object
                var data = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"result", result}, {"roleId", roleId}, {"roleName", roleName}};
                return data;
            }
            case 10003:
            {
                // 结果
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = encoding.GetString(reader.ReadBytes(dataLength));
                return data;
            }
            case 10004:
            {
                // 结果
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = encoding.GetString(reader.ReadBytes(dataLength));
                return data;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }
}